File: backend/src/main/java/at/ac/tuwien/sepm/groupphase/backend/service/AddressService.java
package at.ac.tuwien.sepm.groupphase.backend.service;

import java.util.List;

public interface AddressService {

    /**
     * Get all distinct Countries that are in the database.
     *
     * @return ordered list of all unique countries.
     */
    List<String> getUniqueCountries();

    /**
     * Get all distinct Postal Codes that are in the database.
     *
     * @return ordered list of all unique postal codes.
     */
    List<String> getUniquePostalCodes();

    /**
     * Get all distinct Cities that are in the database.
     *
     * @return ordered list of all unique cities.
     */
    List<String> getUniqueCities();

}

File: frontend/src/app/services/address.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {Globals} from '../global/globals';
import {Observable} from 'rxjs';
import {Address} from '../dtos/Address';


@Injectable({
  providedIn: 'root'
})
export class AddressService {

  private authBaseUri: string = this.globals.backendUri + '/addresses';

  constructor(private httpClient: HttpClient, private globals: Globals) {
  }

  /**
   * Persists Customer to the backend
   *
   * @param address address to add
   */
  createAddress(address: Address): Observable<Address> {
    console.log('Create address ' + address.toString());
    return this.httpClient.post(this.authBaseUri, address);
  }

  /**
   * Get all unique countries from the backend
   *
   */
  getUniqueCountries(): Observable<Address> {
    console.log('Get unique countries ');
    return this.httpClient.get(this.authBaseUri + '/countries');
  }

  /**
   * Get all unique postal codes from the backend
   *
   */
  getUniquePostalCodes(): Observable<Address> {
    console.log('Get unique postal codes ');
    return this.httpClient.get(this.authBaseUri + '/codes');
  }

  /**
   * Get all unique cities from the backend
   *
   */
  getUniqueCities(): Observable<Address> {
    console.log('Get unique cities');
    return this.httpClient.get(this.authBaseUri + '/cities');
  }
}
